### ISSUE TITLE
generated Builder code for generics doesn't use actual types (unlike the value class)

### ISSUE THREAD

**Issue by dsimmsatsquare** [google/auto#676]
Hi, thanks for auto-value!  I ran into something surprising that I think ought to work but chokes in an interesting way (to me anyway).

Let's say I have an abstract class that's generic, like so:

```java
public abstract class SomeThing<E> {
  public abstract int id();
  public abstract NonCollectionNonAutoValueGenericValue<E> someValue();

  public abstract static class Builder<E, T extends SomeThing<E>, B extends Builder<E, T, B>> {
    public abstract B id(int id);
    public abstract B someValue(NonCollectionNonAutoValueGenericValue<E> someValue);
    public abstract T build();
  }
}
```

...and an autovalue class that extends that:

```java
import com.google.auto.value.AutoValue;

@AutoValue public abstract class StringThing extends SomeThing<String> {

  public static Builder newBuilder() {
    return new AutoValue_StringThing.Builder();
  }

  @AutoValue.Builder public abstract static class Builder
      extends SomeThing.Builder<String, StringThing, StringThing.Builder> { }
}
```

.  The generated code for the builder doesn't compile:

```text
$ ./gradlew jar

> Task :compileJava FAILED
/Users/dsimms/Development/thirdparty/auto-value-bug/build/classes/java/main/AutoValue_StringThing.java:71: error: cannot find symbol
    public StringThing.Builder someValue(NonCollectionNonAutoValueGenericValue<E> someValue) {
                                                                               ^
  symbol:   class E
  location: class Builder
1 error

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
1 actionable task: 1 executed
```

Notice that the getter someValue returns `NonCollectionNonAutoValueGenericValue<String>` but the someSet setter in the builder accepts `NonCollectionNonAutoValueGenericValue<E>` and there's no E type parameter any where in sight:

```java
import javax.annotation.processing.Generated;

@Generated("com.google.auto.value.processor.AutoValueProcessor")
final class AutoValue_StringThing extends StringThing {

  private final int id;

  private final NonCollectionNonAutoValueGenericValue<String> someValue;

  private AutoValue_StringThing(
      int id,
      NonCollectionNonAutoValueGenericValue<String> someValue) {
    this.id = id;
    this.someValue = someValue;
  }

  @Override
  public int id() {
    return id;
  }

  @Override
  public NonCollectionNonAutoValueGenericValue<String> someValue() {
    return someValue;
  }

  @Override
  public String toString() {
    return "StringThing{"
         + "id=" + id + ", "
         + "someValue=" + someValue
        + "}";
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (o instanceof StringThing) {
      StringThing that = (StringThing) o;
      return (this.id == that.id())
           && (this.someValue.equals(that.someValue()));
    }
    return false;
  }

  @Override
  public int hashCode() {
    int h$ = 1;
    h$ *= 1000003;
    h$ ^= id;
    h$ *= 1000003;
    h$ ^= someValue.hashCode();
    return h$;
  }

  static final class Builder extends StringThing.Builder {
    private Integer id;
    private NonCollectionNonAutoValueGenericValue<String> someValue;
    Builder() {
    }
    @Override
    public StringThing.Builder id(int id) {
      this.id = id;
      return this;
    }
    @Override
    public StringThing.Builder someValue(NonCollectionNonAutoValueGenericValue<E> someValue) {
      if (someValue == null) {
        throw new NullPointerException("Null someValue");
      }
      this.someValue = someValue;
      return this;
    }
    @Override
    public StringThing build() {
      String missing = "";
      if (this.id == null) {
        missing += " id";
      }
      if (this.someValue == null) {
        missing += " someValue";
      }
      if (!missing.isEmpty()) {
        throw new IllegalStateException("Missing required properties:" + missing);
      }
      return new AutoValue_StringThing(
          this.id,
          this.someValue);
    }
  }

}
```

.  The annotation processor really is working very hard to make sure they line up, like if I even change the *name* of the type parameter in SomeThing.Builder from E to BE (for example):

```java 
public abstract class SomeThing<E> {
  public abstract int id();
  public abstract NonCollectionNonAutoValueGenericValue<E> someValue();

  public abstract static class Builder<BE, T extends SomeThing<BE>, B extends Builder<BE, T, B>> {
    public abstract B id(int id);
    public abstract B someValue(NonCollectionNonAutoValueGenericValue<BE> someValue);
    public abstract T build();
  }
}
```

...then: 

```text
~/Development/thirdparty/auto-value-bug ./gradlew jar

> Task :compileJava FAILED
/Users/dsimms/Development/thirdparty/auto-value-bug/src/main/java/SomeThing.java:7: error: Parameter type NonCollectionNonAutoValueGenericValue<BE> of setter method should be NonCollectionNonAutoValueGenericValue<E> to match getter StringThing.someValue
    public abstract B someValue(NonCollectionNonAutoValueGenericValue<BE> someValue);
                      ^
1 error

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
1 actionable task: 1 executed
```

...or other actual type changes, as usual.  So the processor surely knows what's what there but isn't generating the code with it.   

And this class is just something that's not a collection or another autovalue class, since there's a quite a bit of magic in the annotation processor I know and wanted to be sure it wasn't related to that

```java
import java.util.Objects;

public final class NonCollectionNonAutoValueGenericValue<T> {
  private final T value;

  private NonCollectionNonAutoValueGenericValue(T value) {
    this.value = value;
  }

  public T getValue() {
    return value;
  }

  public static <T> NonCollectionNonAutoValueGenericValue<T> create(T value) {
    return new NonCollectionNonAutoValueGenericValue<>(value);
  }

  // yeah, yeah, this is what autovalue is for, I know.
  @Override public int hashCode() {
    return value.hashCode();
  }

  @Override public boolean equals(Object that) {
    if (!(that instanceof NonCollectionNonAutoValueGenericValue)) {
      return false;
    }
    // noinspection unchecked
    return Objects.equals(getValue(), ((NonCollectionNonAutoValueGenericValue<T>)that).getValue());
  }
}
```

...and my dumb self-contained repro in zip for:

[auto-value-bug.zip](https://github.com/google/auto/files/2578638/auto-value-bug.zip)

Thanks for your help!

**Comment by eamonnmcmanus** [comment]
There's definitely room for improvement here. A lot of the difficulty stems from there being two sets of type parameters in the general case, one for the `@AutoValue` class and one for the `@AutoValue.Builder` class. We try to make sure that they look the same, but they are not *actually* the same. And in your example they are wildly different. :-)

The right solution is probably appropriate use of [`Types.asMemberOf`](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/util/Types.html#asMemberOf-javax.lang.model.type.DeclaredType-javax.lang.model.element.Element-) to determine the actual signature of the abstract methods that the Builder implementation is overriding. In the past we've shied away from using `Types.asMemberOf` because historically it has been very broken in the Eclipse compiler. [This bug](https://bugs.eclipse.org/bugs/show_bug.cgi?id=499026) is still open, which makes me think that it still is broken.

**Comment by dsimmsatsquare** [comment]
First, I should say that the processor does really a pretty good job of making sure the setters in the builder and getters in the value class line up.  I fiddled with these a bunch trying to see if the annotation processor would accept something that didn't make sense or wouldn't accept something that did, but couldn't (except for the silly type parameter name on which doesn't matter very much). 

But back to the generated code: the generated code for the value class does the right thing (ends up being typed with the actual type parameter), so I wonder about how those are different than the setter in the builder (which has the same type parameterization, even if they could be different).  

I can hear you suggesting UTS, which I'll do in some free time...

**Comment by eamonnmcmanus** [comment]
This should be fixed in 1.6.4, which has just been released.

**Comment by dsimmsatsquare** [comment]
Awesome, thanks @eamonnmcmanus !
